interop/tapir/src/main/scala/caliban/interop/tapir/TapirAdapter.scala

```diff
@@ -0,0 +1,264 @@
+  implicit val wsOutputSchema: Schema[GraphQLWSOutput]               = Schema.derivedSchema
+
+  def makeHttpService[R, E](
+    interpreter: GraphQLInterpreter[R, E],
+    skipValidation: Boolean = false,
+    enableIntrospection: Boolean = true,
+    queryExecution: QueryExecution = QueryExecution.Parallel
+  )(implicit
+    requestCodec: JsonCodec[GraphQLRequest],
+    responseCodec: JsonCodec[GraphQLResponse[E]]
+  ): List[ServerEndpoint[GraphQLRequest, Unit, GraphQLResponse[E], Any, RIO[R, *]]] = {
+
+    def queryFromQueryParams(queryParams: QueryParams): DecodeResult[GraphQLRequest] =
+      for {
+        req <- requestCodec.decode(s"""{"query":"","variables":${queryParams
+                 .get("variables")
+                 .getOrElse("null")},"extensions":${queryParams
+                 .get("extensions")
+                 .getOrElse("null")}}""")
+
+      } yield req.copy(query = queryParams.get("query"), operationName = queryParams.get("operationName"))
+
+    val postEndpoint: Endpoint[GraphQLRequest, Unit, GraphQLResponse[E], Any] =
+      endpoint.post
+        .in(
+          (headers and stringBody and queryParams).mapDecode { case (headers, body, params) =>
+            val getRequest =
+              if (params.get("query").isDefined)
+                queryFromQueryParams(params)
+              else if (headers.contains(Header("content/type", "application/graphql")))
+                DecodeResult.Value(GraphQLRequest(query = Some(body)))
+              else requestCodec.decode(body)
+
+            getRequest.map(request =>
+              headers
+                .find(r => r.name == GraphQLRequest.`apollo-federation-include-trace` && r.value == GraphQLRequest.ftv1)
+                .fold(request)(_ => request.withFederatedTracing)
+            )
+          }(request => (Nil, requestCodec.encode(request), QueryParams()))
+        )
+        .out(customJsonBody[GraphQLResponse[E]])
+
+    val getEndpoint: Endpoint[GraphQLRequest, Unit, GraphQLResponse[E], Any] =
+      endpoint.get
+        .in(
+          queryParams.mapDecode(queryFromQueryParams)(request =>
+            QueryParams.fromMap(
+              Map(
+                "query"         -> request.query.getOrElse(""),
+                "operationName" -> request.operationName.getOrElse(""),
+                "variables"     -> request.variables
+                  .map(_.map { case (k, v) => s""""$k":${v.toInputString}""" }.mkString("{", ",", "}"))
+                  .getOrElse(""),
+                "extensions"    -> request.extensions
+                  .map(_.map { case (k, v) => s""""$k":${v.toInputString}""" }.mkString("{", ",", "}"))
+                  .getOrElse("")
+              ).filter { case (_, v) => v.nonEmpty }
+            )
+          )
+        )
+        .out(customJsonBody[GraphQLResponse[E]])
+
+    def logic(request: GraphQLRequest): RIO[R, Either[Unit, GraphQLResponse[E]]] =
```

Comment:
What is the `Unit` for?

Reply:
This is the tapir default error type. Since errors are already included in `GraphQLResponse`, I didn't change it.

Reply:
Actually I'm looking into porting `ContextWrapper` now and I might change `Unit` to `StatusCode` to allow the wrapper failing with a status code.